Peernotifier: Add Peer Notifier package

carlaKC opened the pull request:
This PR adds a peer notifier package which supplies clients with peer online/offline subscriptions and a corresponding rpc endpoint. 

It is intended for use in work related to #3332 to incorporate peer online/offline events in channel scoring. 

#### Pull Request Checklist
- [x] All changes are Go version 1.12 compliant, commented and formatted
- [x] Any new logging statements use an appropriate subsystem and
  logging level - peer logger was used, it felt like the most appropriate?
- [x] Commits build properly, tests pass
- [x] Make check, go vet and make lint ok

Roasbeef commented:
If it's for #3332, then why do we need the RPC end point as well?

carlaKC commented:
> If it's for #3332, then why do we need the RPC end point as well?

I thought that it would be a useful endpoint to have for people writing their own channel management logic? Eg. somebody who's not using autopilot, and is deciding which one of their peers to open a channel to. If they subscribe to peer events and store them, they'd have an idea of how reliably online each peer has been.

halseth commented on peernotifier/log.go:
```diff
@@ -0,0 +1,29 @@
+
+// The default amount of logging is none.
+func init() {
+	UseLogger(build.NewSubLogger("PEER", nil))
```
I think `PEER` is most likely already in use?

carlaKC replied:
It is, but from what I can tell it won't be used at all except in unit tests because the `genSubLogger` param is nil? 

More generally, I wasn't sure whether to make a new sublogger in `lnd/log.go` for peer notifier (it felt a bit much since it's only used a handful of times) so I reused the peer logger. The comments in `lnd/log.go` say new logger "per subsystem"?

cfromknecht replied:
i'm okay with using the new `PRNF` tag for now, we can consolidate later if need be

halseth commented on peernotifier/peernotifier.go:
```diff
@@ -0,0 +1,94 @@
+func (p *PeerNotifier) NotifyPeerOnline(pubKey *btcec.PublicKey) {
+	event := PeerOnlineEvent{PubKey: pubKey}
+
+	log.Debugf(
```
style nit: For logging and other string formatting calls we usually do
```go
log.Debugf("PeerNotifier notifying peer: %x online",
		string(pubKey.SerializeCompressed()[:]))
```

carlaKC replied:
cool, so generally for string formatting, keep the string inline and then newlines for all the vars like you do for functions? 
```
log.Debugf("Long sentence taking up space: %x and two values: %v", 
                   firstVar, 
                   secondVar,
)
```

Or just wrap at 80
```
log.Debugf("Long sentence taking up space: %x and two values: %v", 
                   firstVar, secondVar)
```

cfromknecht replied:
@carlaKC typically we shoot for the latter wrt to logging statements

carlaKC commented:
Ok great, updated to fix style nit and remove rpc code (in favour of a separate PR). 

I've changed the logger to still use `peerLog`, just with a different tag, although I'm not sure whether `peerNotifier` should use `peerLog`, `ntfnLog` or a new logger?

cfromknecht commented on peernotifier/peernotifier.go:
```diff
@@ -0,0 +1,92 @@
+	event := PeerOnlineEvent{PubKey: pubKey}
+
+	log.Debugf("PeerNotifier notifying peer: %x online",
+		string(pubKey.SerializeCompressed()[:]),
```
the argument here i believe only needs to be `pubKey.SerializeCompressed()`. also one of the few exceptions to our line wrapping is for log statements, so this would be acceptable:
```golang
        log.Debugf("PeerNotifier notifying peer: %x online",
                        pubKey.SerializeCompressed())
```

cfromknecht commented on peernotifier/peernotifier.go:
```diff
@@ -0,0 +1,92 @@
+// PeerOnlineEvent represents a new event where a peer comes online.
+type PeerOnlineEvent struct {
+	// PubKey is the ecdsa.PubKey of the peer that came online.
+	PubKey *btcec.PublicKey
```
we may want to consider making these just `[33]byte` arrays, since for the most i think we only need to use this as an identifier. not a block change though, it's something we can consider in the future

carlaKC replied:
Agreed that it's more likely to be used as an identifier, so I've changed to `[33]byte`. Extra info can always be added if it's needed, but I don't see much of a case for it as is.

carlaKC commented:
Rebased and updated to fix formatting nits & switch pubkeys to `[33] byte` since they'll most likely just be used as identifiers. 
PTAL @halseth 🙏

halseth commented on log.go:
```diff
@@ -119,6 +120,7 @@ func init() {
 	chanbackup.UseLogger(chbuLog)
 	monitoring.UseLogger(promLog)
 	wtclient.UseLogger(wtclLog)
+	peernotifier.UseLogger(peerLog)
```
You need to create a new subsystem logger above, with tag "PRNF". If you look at the `subsystemsLoggers` map below, there should be a 1:1 mapping between tags and loggers :)

carlaKC replied:
Ok cool, I was wondering about that - `ntfnLog` threw me off a bit because it's used twice for `NTFN` and `NTFR`? Typo maybe?

halseth replied:
Ooh, yeah that's a typo. I put up a fix: https://github.com/lightningnetwork/lnd/pull/3374

Good catch! 😅

halseth commented on peernotifier/peernotifier.go:
```diff
@@ -0,0 +1,87 @@
+// PeerOnlineEvent represents a new event where a peer comes online.
+type PeerOnlineEvent struct {
+	// PubKey is the peer's compressed public key.
+	PubKey [33]byte
```
+1 on this change! 👍

halseth commented on server.go:
```diff
@@ -2722,6 +2734,13 @@ func (s *server) addPeer(p *peer) {
+	// Inform the peer notifier of a peer online event so that it can be reported
+	// to clients listening for peer events.
+	var pubKey [33]byte
+	copy(pubKey[:], p.addr.IdentityKey.SerializeCompressed())
```
optimization: do this before creating the `pubStr` above, so don't we have to serialize twice :)

halseth commented on server.go:
```diff
@@ -2972,6 +2991,13 @@ func (s *server) removePeer(p *peer) {
+	// Inform the peer notifier of a peer offline event so that it can be
+	// reported to clients listening for peer events.
+	var pubKey [33]byte
+	copy(pubKey[:], p.addr.IdentityKey.SerializeCompressed())
```
same here
